refactor: moving update logic to a separate file

axel7083 opened the pull request:
### What does this PR do?

Moving the logic of the updater system (Status Bar Item, error handling etc.) from the index to a dedicated file.

### Screenshot / video of UI

<!-- If this PR is changing UI, please include
screenshots or screencasts showing the difference -->

### What issues does this PR fix or reference?

Related to https://github.com/containers/podman-desktop/issues/5754

### How to test this PR?

<!-- Please explain steps to verify the functionality,
do not forget to provide unit/component tests -->

- [x] Tests are covering the bug fix or the new feature

benoitf commented on packages/main/src/plugin/updater.spec.ts:
```diff
@@ -0,0 +1,165 @@
+import { UPDATER_UPDATE_AVAILABLE_ICON } from '/@/plugin/index.js';
+import { Updater } from '/@/plugin/updater.js';
+
+const mocks = vi.hoisted(() => ({
+  // MessageBoxMock
+  showMessageBoxMock: vi.fn(),
+  // StatusBarRegistry
+  setEntryMock: vi.fn(),
+  // CommandRegistry
+  registerCommandMock: vi.fn(),
+  executeCommandMock: vi.fn(),
+  // Electron App
+  getVersionMock: vi.fn(),
+  // Electron Update
+  downloadUpdateMock: vi.fn(),
+  quitAndInstallMock: vi.fn(),
+  checkForUpdatesMock: vi.fn(),
+  onMock: vi.fn(),
+  // utils
+  isLinuxMock: vi.fn(),
+}));
```
hello, I'm not sure to understand why we need this block ?

if you do
```ts
vi.mock('electron-updater', () => ({
  autoUpdater: {
    downloadUpdate: vi.fn()
})
```

benoitf replied:
and then `vi.mocked(...)` to access the mocked method

axel7083 replied:
I have a preference for this method, of declaring all the mocks in a single places, and accessing them through `mocks.[name]`

We see often some `vi.spy`, of `vi.mock` but when we want to set default implementation in the beforeEach combined with a `vi.resetAllMocks();` it is easier to access those mock isntance through a single place, rather than calling several methods.

There are in the code, also some tests declaring `MockInstance` objects, in multiple places, in AI-Studio we were in favour of using this single mocks objects to keep them in a single places

benoitf replied:
they'll still be declared at a single place

currently you're losing type by doing that

benoitf reviewed (changes requested):
please review the mocking strategy to have something with typesafety

axel7083 commented:
> please review the mocking strategy to have something with typesafety

@benoitf done

benoitf commented on packages/main/src/plugin/updater.spec.ts:
```diff
@@ -0,0 +1,156 @@
+
+  // Simulate PROD env
+  vi.stubEnv('PROD', 'true');
+
+  vi.spyOn(app, 'getVersion').mockReturnValue('@debug');
+  // eslint-disable-next-line no-null/no-null
+  vi.spyOn(autoUpdater, 'checkForUpdates').mockResolvedValue(null);
+
+  vi.spyOn(commandRegistry, 'executeCommand').mockResolvedValue(undefined);
+  vi.spyOn(util, 'isLinux').mockReturnValue(false);
```
note: as it's mocked you can avoid the spy and do vi.mocked(app.getVersion).mockReturnValue('@debug');

axel7083 replied:
done

benoitf commented on packages/main/src/plugin/updater.spec.ts:
```diff
@@ -0,0 +1,156 @@
+  const updater = new Updater(messageBoxMock, statusBarRegistryMock, commandRegistry);
+  const disposable: unknown = updater.init();
+  expect(disposable).toBeDefined();
+  expect(disposable instanceof Disposable).toBe(true);
```
```suggestion
  expect(disposable instanceof Disposable).toBeTruthy();
```

benoitf commented on packages/main/src/plugin/updater.spec.ts:
```diff
@@ -0,0 +1,156 @@
+      expect(iconClass).toBe(UPDATER_UPDATE_AVAILABLE_ICON);
+      expect(enabled).toBe(true);
+      expect(command).toBe('update');
+      expect(highlight).toBe(true);
```
```suggestion
      expect(highlight).toBeTruthy();
```

benoitf commented on packages/main/src/plugin/updater.spec.ts:
```diff
@@ -0,0 +1,156 @@
+      expect(text).toBe('v@debug');
+      expect(tooltip).toBe('Update available');
+      expect(iconClass).toBe(UPDATER_UPDATE_AVAILABLE_ICON);
+      expect(enabled).toBe(true);
```
```suggestion
      expect(enabled).toBeTruthy();
```

benoitf reviewed (commented):
thanks for the update

I think you don't need the spy as they're already mocked

I'll try the PR

benoitf commented:
I tried but unfortunately there is an error

https://github.com/containers/podman-desktop/assets/436777/aa41dcb8-623b-4870-9191-70f268e22def

> How I tested ?
I replaced package.json version with 1.6.5 and did `yarn compile:current` and launched the dmg (and it's telling me there is an update from 1.6.5 to 1.7.1)
I click on proceed but then I have a javascript error


Tried with the main branch I got a real error message (saying that my volume was read-only)

axel7083 commented:
@benoitf thanks for pointing out this issue, the tests I've added were not covering properly the error scenarios. The `error` listener was not properly bind to the current class, leading to the javascript error you encountered. 

I fixed the issue and added a regression test.

benoitf commented:
thanks @axel7083 

let me check again

benoitf reviewed (approved):
did the error as expected and successful update when not using read-only volume 🎉 
![Screenshot 2024-02-28 at 10 49 48](https://github.com/containers/podman-desktop/assets/436777/a18da4c2-5311-4e8a-81e9-9d14256ee335)

benoitf commented on packages/main/src/plugin/updater.ts:
```diff
@@ -0,0 +1,265 @@
+  /**
+   * Sets the default version entry in the status bar.
+   */
+  private defaultVersionEntry() {
```
it looks like we're missing return type

benoitf commented on packages/main/src/plugin/updater.ts:
```diff
@@ -0,0 +1,265 @@
+  /**
+   * Sets the update available entry in the status bar.
+   */
+  private updateAvailableEntry() {
```
it looks like we're missing return type

benoitf commented on packages/main/src/plugin/updater.ts:
```diff
@@ -0,0 +1,265 @@
+  /**
+   * Registers commands related to version and update.
+   */
+  private registerCommands() {
```
it looks like we're missing return type

axel7083 replied:
Why is the linter not saying something ?

benoitf replied:
probably something is missing 🤔

benoitf replied:
https://github.com/containers/podman-desktop/pull/6183
